doc/rm/c_cpp_coding_style.md

```diff
@@ -173,6 +186,49 @@ This may change if we find cases where this initialization improves readability.
 When initializing an array, initializers *may* be designated when that makes the array more readable (e.g., lookup tables that are mostly zeroed). Mixing designated and undesignated initializers, or using nested initializers, is still
 forbidden.
 
+### Nonstandard Attributes
+
+The following nonstandard attributes may be used:
+*   `section(<name>)` to put a definition into a given object section.
+*   `weak` to make a symbol definition have weak linkage.
+*   `naked` to ensure a function has no prolog or epilog.
+*   `packed` to ensure a struct contains no padding.
+
+It is recommended that other nonstandard attributes are not used.
+
+All nonstandard attributes must be supported by both GCC and Clang.
+
+***The use of nonstandard attributes must be hidden by a macro definition starting `OT_`.***
+
+For instance, `__attribute__((section("foo")))` should not be used on a declaration.
+Instead, the attribute should be hidden behind a macro, like below.
+```c
+#define OT_SECTION(name) __attribute__((section(name)))
+//...
+void _crt(void) OT_SECTION(".crt");
+```
+
+This ensures changes to add compatibilty for other compilers are less invasive.
+
+### Nonstandard Compiler Builtins
+
+All nonstandard builtins must be supported by both GCC and Clang.
+
+***The use of nonstandard compiler builtins must be hidden by a compatible definition.***
+
+For instance, `__builtin_abort()` should not be called.
+Instead, the call should be hidden by a compatible definition, like below.
+```c
+_Noreturn void abort(void) {
+  __builtin_abort();
+}
+```
+
+This ensures changes to add compatibilty for other compilers are less invasive.
+
+To avoid overhead, this function can be declared as an `inline` function in a header file.
+If it is declared as an inline function, exactly one compilation unit must have a compatible `extern inline` declaration for that function.
```

Comment:
Maybe include an example? C `inline` tends to be really finicky to get right.

Reply:
Will do.